envoy/tcp/async_tcp_client.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/connection.h"
+
+namespace Envoy {
+namespace Tcp {
+
+/*
+ * AsyncTcpClientCallbacks for peer connection callbacks and data.
+ */
+class AsyncTcpClientCallbacks : public Network::ConnectionCallbacks {
+public:
+  ~AsyncTcpClientCallbacks() override = default;
+
+  /*
+   * Invoked when data is delivered from the peer connection.
+   * @param data supplies data from the upstream
+   * @param end_stream whether the data is the last data frame
+   */
+  virtual void onData(Buffer::Instance& data, bool end_stream) PURE;
+};
+
+/**
+ * An async TCP client with connection to peers.
+ */
+class AsyncTcpClient {
+public:
+  virtual ~AsyncTcpClient() = default;
+
+  /**
+   * Connect to a remote host. Errors or connection events are reported via the
+   * event callback registered via addAsyncTcpClientCallbacks().
+   */
+  virtual bool connect() PURE;
+
+  /**
+   * Close the client, the underlying connection will also be defer deleted when
+   * a Close event is received.
+   */
+  virtual void close() PURE;
```

Comment:
The semantics of the call need to be clarified a little. Presently it abortively closes the connection discarding any unsent data.

Reply:
Done